### DISCUSSION TITLE
Modification - Mobile Breakpoint

### DISCUSSION THREAD

**Discussion by james-hayden-ellis** [Legal-and-General/canopy#789] [Canopy requests]
**Type of request**
Modification

**Component, pattern or foundation?**
Foundation

**Nature of request**
Enquiring to see whether we should use the recoding of the DOTCOM front end as a way to update our mobile breakpoint (currently 320px) and what that breakpoint should be.

General UK stats for the past 12 months : https://gs.statcounter.com/screen-resolution-stats/mobile/united-kingdom

<img width="1439" alt="Screenshot 2022-03-28 at 15 18 02" src="https://user-images.githubusercontent.com/59871034/160418460-f8e5a9bc-65c2-4524-acc8-1914466e14fa.png">
<img width="1440" alt="Screenshot 2022-03-28 at 15 18 24" src="https://user-images.githubusercontent.com/59871034/160418505-93d600ad-227d-42cd-9ff5-6b906dafa65e.png">

L&G stats are a bit skewed as according to Mike Tomsett: "There is a know issue with iPhones where they don't properly report the user agent and therefore we don't get completely accurate device information beyond "it's an iPhone"! and there's a default setting in Adobe for that viewport. Matt's (Blake) recommendation was to exclude iPhones from the results and see what they look like after that. They could look at a better solution for getting that information into Adobe, but it won't be quick enough for us to make the decisions we need."

But here are the Adobe Analytics for mobile screens for the last 12 months (though I can't figure out how to exclude the 320x480px)...

![Adobe Analytics Mobile View April 2021-2022](https://user-images.githubusercontent.com/59871034/161945309-ba804e23-bcfd-4d5c-8d7f-0f81d3f03a5c.png)

**Comment by alexcanning** [comment]
Have we landed on a safe / most commonly used minimum width?

**Comment by owensgit** [comment]
I think it's worth revisiting the `sm` breakpoint (320px), in some ways it's not that helpful for us, because when styling something at `min-width: 320px`, we're essentially saying that this is the base style for lowest screenwidth possible, so no need to use a breakpoint at all. The same goes for `max-width: 320px`, the user would never see the styles because they will never view it on a device less than 320px.

I would sugest that the `sm` breakpoint could be updated to we can add modifications for users with large mobiles, which aren't quite tablets - small enough to need a mobile layout, but large enough to do more with the available space than a smaller phone.

If so it would need to sit somewhere between 0 and 768px (`md` breakpoint).

For reference, here are our current breakpoints:

```
$breakpoints: (
  // 320px
  'sm': 20rem,
  // 768px
  'md': 48rem,
  // 1024px
  'lg': 64rem,
  // 1280px
  'xl': 80rem,
  // 1440px
  'xxl': 90rem
) !default;

```

**Reply by elenagarrone** [reply]
@owensgit I agree with you 👍

**Reply by alexcanning** [reply]
Could we introduce a new XS breakpoint which has a max-width of less than SM, and make a SM min-width slightly more useful?

**Reply by elenagarrone** [reply]
Yes, that would be possible but is it really needed? Who is doing this investigation? will it be dotocom?

**Reply by james-hayden-ellis** [reply]
Here's the original JIRA ticket request...

<div class="qwux5g-0 ioAami" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0px; padding: 0px; display: flex; -webkit-box-align: center; align-items: center; flex-wrap: nowrap; -webkit-box-pack: justify; justify-content: space-between;"><div class="qwux5g-1 iKmxyr" style="margin: 0px; padding: 0px; display: flex; -webkit-box-pack: start; justify-content: flex-start; -webkit-box-align: center; align-items: center; max-width: 100%;"><h2 class="qwux5g-2 dqHXQe" style="margin: 0px; padding: 0px; font-style: inherit; color: var(--ds-text,#172B4D); letter-spacing: -0.003em; font-size: 1em; line-height: 24px; font-weight: 600; text-transform: none; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; overflow-wrap: break-word; min-width: 0px;">Description</h2></div></div><div role="presentation" data-test-id="issue.views.field.rich-text.description" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none; margin: 0px; padding: 0px;"><div class="sc-1oys0dl-0 fuBDPm" style="margin: 0px 0px 24px; padding: 0px;"><div class="RootWrapper-sc-1va80k6-0 fsVJT" style="margin: 0px 0px 0px -8px; padding: 0px; flex-direction: column; width: 711.359375px; display: flex; flex: 0 1 auto; max-width: 100%;"><div style="margin: 0px; padding: 0px; position: absolute;"><label class="Label__LabelWrapper-sc-17towfw-0 keOXhT" style="display: block;"><div class="ContentWrapper-kdagst-0 fPLsyX" style="margin: 0px; padding: 0px; max-width: 100%; position: relative;"><div class="FieldBaseWrapper-sc-14kmybr-0 cspqsP" style="margin: 0px; padding: 0px; display: block; max-width: 100%; flex: 1 1 auto; -webkit-box-flex: 1;"><div class="Content__ContentWrapper-ve26fj-2 feWTNh" style="margin: 0px; padding: 0px; flex: 0 0 100%; display: block; max-width: 100%;"><div class="Content__ChildWrapper-ve26fj-0 ilJUcG" style="margin: 0px; padding: 0px; display: flex; flex-direction: column; flex: 1 0 auto; max-width: 100%; min-height: 32px !important;"><div class="Content-ve26fj-1 llCsQv" style="margin: 0px; padding: 0px 6px; -webkit-box-align: center; align-items: center; background-color: transparent; border: 2px solid transparent; border-top-left-radius: 3px; border-top-right-radius: 3px; border-bottom-right-radius: 3px; border-bottom-left-radius: 3px; box-sizing: border-box; color: rgb(9, 30, 66); display: flex; flex: 1 0 auto; font-size: 14px; -webkit-box-pack: justify; justify-content: space-between; line-height: 1.4285714285714286; max-width: 100%; overflow: hidden; transition: background-color 0.2s ease-in-out, border-color 0.2s ease-in-out; word-wrap: break-word;"><div class="ReadViewContentWrapper-xymwx5-0 kLiHRY" style="margin: 0px; padding: 0px; display: flex; flex: 1 1 auto; min-width: 0px; width: 695.359375px;"><div class="sc-9qk4vv-0 gHBtpa" style="margin: 0px; padding: 0px; width: 695.359375px;"><div style="margin: 0px; padding: 0px;"><div role="presentation" style="margin: 0px; padding: 0px;"><div class="ak-renderer-wrapper sc-1cb0syy-0 eibAuu" style="margin: 0px; padding: 0px; position: relative; width: 695.359375px;"><div style="margin: 0px; padding: 0px; display: block; width: 695.359375px; position: absolute;"></div><div class="sc-1u3iq5d-0 dwYBpi" style="margin: 0px; padding: 0px; font-size: 14px; line-height: 1.5rem; color: rgb(23, 43, 77); word-wrap: break-word; white-space: pre-wrap;"><div class="ak-renderer-document" style="margin: 0px; padding: 0px;"><p data-renderer-start-pos="1" style="margin: 0px; padding: 0px; font-size: 1em; line-height: 1.714; font-weight: normal; letter-spacing: -0.005em;">Currently the responsive image block automatically creates 3 smaller versions of an original uploaded image as below.<br>These are also set as the breakpoints for viewport sizes of each image.</p><div class="pm-table-container  drwdyq-0 dugMha" data-layout="default" style="margin: 0px auto 16px; padding: 0px; position: relative; box-sizing: border-box; z-index: 0; transition: all 0.1s linear; display: flex; clear: both; width: inherit;"><div class="pm-table-wrapper" style="margin: 0px; padding: 0px; overflow-x: auto;">

Name | Image width (px) | breakpoints (px)
-- | -- | --
Original (large desktop) | 1400 - 2000 | >=1013
Desktop | 1012 | 769=<=1012
Tablet | 768 | 321=<=768
Mobile | 320 | <=320

</div></div><p data-renderer-start-pos="394" style="margin: 0.75rem 0px 0px; padding: 0px; font-size: 1em; line-height: 1.714; font-weight: normal; letter-spacing: -0.005em;">Should we continue to use these image sizes / breakpoints?<br>If new image sizes / breakpoints are needed, what should they be?</p><p data-renderer-start-pos="520" style="margin: 0.75rem 0px 0px; padding: 0px; font-size: 1em; line-height: 1.714; font-weight: normal; letter-spacing: -0.005em;">Particular attention should be paid to mobile. Most mobile viewports these days appear to be around 414px wide meaning most devices load the tablet image version.</p></div></div></div></div></div></div></div></div></div></div></div></div></label></div></div></div></div>

**Reply by owensgit** [reply]
Having a look at [Bootstrap's most recent breakpoint values](https://getbootstrap.com/docs/5.1/layout/breakpoints/), they have a `xs` which is actually just `0` - essentially not a breakpoint, but just the default mobile-first setting. And then they have a `sm` which is `576px`. This is similar to @alexcanning 's suggestion, but I'd recommend not having `xs`, as @elenagarrone also suggested, as it seems kind of pointless - why would we do `min-width: 0` or `max-width: 0`? 😄   

[Tailwind CSS ](https://tailwindcss.com/docs/responsive-design) does this, it has a `sm` breakpoint of `640px`, but no `xs`.

My favoured approach would be to increase our `sm` breakpoint to either `576px` or `640px` - whichever one starts to look a bit stretched on larger mobiles/small tablets, therefore opening the door ot creating some nicer layouts/optimisations in that region. It will need thorough testing of existing components and features that work with responsive styles such as `showAt`, and the responsive margin/padding directives. I guess would also be breaking change. Worth it though :)

@james-hayden-ellis  I think that might work well with the image size issue you meantioned, because the mobile size image could be displayed for devices below `576px` or `640px`, which should be fine - although the image files themselves may need to be optimised for those sizes.

**Comment by james-hayden-ellis** [comment]
Okay, so after reviewing the stats @alexcanning and I settled on suggesting a new 400px viewport size, so that'd roll into and readjust the current system like so...

<!DOCTYPE html>

Name | Image width (px) | breakpoints (px)
-- | -- | --
XS (Small Mobile) | 400px | <=399
SM (Mobile) | 768px | 400 - 767
MD (Tablet) | 1024px | 768 - 1023
LG (Small Desktop) | 1280px | 1024 - 1279
XL (Desktop) | 1440px | 1280 - 1439
XXL (Large Desktop) | 2000px | 1440+

</html>

Anyone (particularly from Dev) see any issues..?

**Reply by owensgit** [reply]
Thanks James, I would say 400px is quite small, do the stats show a lot of devices below 400px? 

Thinking mobile-first and then scaling up, devices below 400px would be the default layout, and then at the 400px breakpoint there's an opportunity to adjust things, but then nothing until 768px. It doesn't seem like much would need to change after 400px. 

I haven't seen the stats, but I'd favour what Bootstrap are doing with 576px. Because a mobile-first layout starts to look stretched at around 576px and needs some adjustments, before jumping into tablet-portrait view at 768px.

The other thing I'd say about the table above is that the breakpoints column aren't mobile-first - e.g. the breakpoints should show the range _above_ that value. E.g..

Name | Image width (px) | breakpoints (px)
-- | -- | --
Default | 0 | 0 - 399
XS (Mobile) | 400px | 400 - 767
SM (Tablet portrait) | 768px | 768 - 1023
_etc..._

Again I think the way [Bootstrap](https://getbootstrap.com/docs/5.1/layout/breakpoints/) have done it is pretty good:

<img width="836" alt="Screenshot 2022-04-26 at 08 20 05" src="https://user-images.githubusercontent.com/1943532/165244188-bfb115b4-2ec3-4811-ad06-7af9bd9933ee.png">

Let me know what you think, happy to jump on a call if it's easier.

Another thing Boostrap do, which I agree with, is that the "MD" breakpoint is actually tablet-portrait (768px) and then large is tablet landscape/small desktop. But may that's another conversation :) Generally I think Boostrap's breakpoints are very good and reflect the diverse range of devices out there and at what point layouts start to look stretched out as you scale up and go through a mobile-first approach.

Naming and other breakpoints aside, if it were totally up to me, I'd go with this... :) 

Name | Image width (px) | breakpoints (px)
-- | -- | --
Default | 0 | > 0
XS (Small Mobile) | 576px | > 576
SM (Mobile) | 768px | > 768
MD (Tablet) | 1024px | > 1024
LG (Small Desktop) | 1280px | > 1280
XL (Desktop) | 1440px | > 1440 
XXL (Large Desktop) | 2000px | > 2000

If I were to totally push the boat out, and play with the names, I'd go with this

Name | Image width (px) | breakpoints (px)
-- | -- | --
Default (Small Mobile)  | 0 | > 0
SM (Mobile) | 576px | > 576
MD (Tablet portrait) | 768px | > 768
LG (Tablet landscape) | 1024px | > 1024
XL (Desktop) | 1280px | > 1280
XXL (Large Desktop) | 1440px | > 1440

**Reply by james-hayden-ellis** [reply]
Hi Owen, thanks for the detailed response, maybe we're looking it from two different viewpoints but when you say...

<!DOCTYPE html>

Name | Image width (px) | breakpoints (px)
-- | -- | --
Default | 0 | 0 - 399
XS (Mobile) | 400px | 400 - 767
SM (Tablet portrait) | 768px | 768 - 1023
etc... |   |  

It sounds like you're stretching (for example) the 400px image to fit any device _up to_ 767px

Anyway, I realise I'd got my sizes out of sync, (and I've re-jigged the image sizes accordingly) it should be...

Name | breakpoints (px) | Image width (px)
-- | -- | --
XS (Small Mobile) | >320px | 320px
SM (Mobile (**NEW**) | >400px | 400px
MD (Tablet Portrait) | >768px | 768px
LG (Tablet Landscape) | >1024px | 1024px
XL (Desktop) | >1280px | 1280px
XXL (Large Desktop) | >1440px | 2000px

Note : Keeping that last 2000px image as an Epi-server legacy thing

**Reply by james-hayden-ellis** [reply]
Could even...

<!DOCTYPE html>

Name | breakpoints (px) | Image width (px)
-- | -- | --
XS (Small Mobile) | >320px | 400px

**Reply by elenagarrone** [reply]
@james-hayden-ellis what would you have from 0 until 320px - no image?

**Reply by james-hayden-ellis** [reply]
@elenagarrone Maybe I'm using the wrong symbols, >320 meaning up to 320.

**Reply by james-hayden-ellis** [reply]
<!DOCTYPE html>

Name | breakpoints (px) | Image width (px)
-- | -- | --
XS (Small Mobile) | ≤320px | 320px (or 400px)
SM (Mobile (NEW) | >321 ≤400px | 400px
MD (Tablet Portrait) | >401 ≤768px | 768px
LG (Tablet Landscape) | >769 ≤1024px | 1024px
XL (Desktop) | >1025 ≤1280px | 1280px
XXL (Large Desktop) | >1281 ≤1440px | 2000px

..?

**Reply by elenagarrone** [reply]
Ok that sounds good 👍  from a dev point of view we wouldn't name the first breakpoint because it's the default style since it's mobile first. Also we should consider the current breakpoints to make this breaking change break less stuff 😅 :

Current breakpoints:
Name | breakpoints (px)
-- | -- 
Default | 0 to 320px 
SM | 321px to 768px 
MD | 769px to 1024px
LG | 1025px to 1280px
XL | 1281px to 1440px
XXL | 1440px up

New proposed breakpoints:
Name | breakpoints (px) | Image width (px)
-- | -- | --
Default | 0 to 320px | 320px (or 400px)
XS | 321px to 400px | 400px
SM | 401px to 768px | 768px
MD | 769px to 1024px | 1024px
LG | 1025px to 1280px | 1280px
XL | 1281px to 1440px | 2000px
XXL | 1440px up | 2000px

**Reply by james-hayden-ellis** [reply]
Okay, so for the purposes of my JIRA ticket we can consider this done and if anything crops up we can revisit it later yeah? 👍🏻

**Reply by elenagarrone** [reply]
@james-hayden-ellis I guess so, but just FYI some work will need to be done to make the changes to the breakpoints in Canopy so that the devs in dotcom can used them to set the image width as per table above (if they are using the Canopy grid).

Not sure if @alexcanning needs to approve too?

**Reply by alexcanning** [reply]
@elenagarrone the XS breakpoint in your table has a very small window of use - propose we lose the XS breakpoint and set `default` to `0 to 399px`, and set `SM` to `400px to 767px`

**Reply by alexcanning** [reply]
We don't seem to have consensus on what the most appropriate breakpoints are at the smaller end of the scale. Do we properly understand the purpose of this exercise?

**Reply by elenagarrone** [reply]
From @owensgit on slack:

> I’d still question if 400px is useful from a mobile-first layout point of view. I can’t see us needing to made any changes at the 400px breakpoint, but I could at 576px for example

**Reply by elenagarrone** [reply]
> @elenagarrone the XS breakpoint in your table has a very small window of use - propose we lose the XS breakpoint and set `default` to `0 to 399px`, and set `SM` to `400px to 767px`

Yes, I agree with you - XS is probably unnecessary. What do we think about what @owensgit has proposed?

**Reply by alexcanning** [reply]
It has merit - allows us to design any small mobile size we like, while giving us provision for larger mobile devices. Could be a good thing. @james-hayden-ellis ?

**Reply by owensgit** [reply]
Don't forget it's the way other frameworks such as Bootstrap and Tailwind have also gone :)

**Reply by elenagarrone** [reply]
@james-hayden-ellis @alexcanning any update on this? I'm currently working on `v8.0.0` of Canopy and it would be good to know if we want to include the changes to the grid. I'm referring in particular to @owensgit's point of the breakpoint at 576px instead of 400px.

**Comment by elenagarrone** [comment]
@alexcanning and I had a chat about modifying the current breakpoints and decided it's not something we want to do for the moment as it needs more investigation.

I'm marking this as resolved and we can pick it up again if an update is needed.
